File: fsharp/yacht/Yacht.fs
module Yacht

type Category =
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One
    | Two
    | Three
    | Four
    | Five
    | Six
    
let private dieScore (die: Die): int =
    match die with
    | One   -> 1
    | Two   -> 2
    | Three -> 3
    | Four  -> 4
    | Five  -> 5
    | Six   -> 6

let private scoreSingleNumber dice goalDie =
    dice
    |> List.filter (fun die -> die = goalDie)
    |> List.sumBy dieScore

let private scoreFullHouse dice =
    match List.countBy id dice with
    | [(_, 2); (_, 3)] -> List.sumBy dieScore dice
    | _ -> 0

let private scoreFourOfAKind dice =
    match List.countBy id dice with
    | [(die, 5)] | [_; (die, 4)] | [(die, 4); _] -> dieScore die * 4
    | _ -> 0

let private scoreLittleStraight dice =
    if List.sort dice = [Die.One; Die.Two; Die.Three; Die.Four; Die.Five] then 30 else 0
    
let private scoreBigStraight dice =
    if List.sort dice = [Die.Two; Die.Three; Die.Four; Die.Five; Die.Six] then 30 else 0
    
let private scoreChoice dice = List.sumBy dieScore dice

let private scoreYacht dice =
    if List.distinct dice |> List.length = 1 then 50 else 0

let score category dice =
    match category with
    | Ones           -> scoreSingleNumber dice Die.One
    | Twos           -> scoreSingleNumber dice Die.Two
    | Threes         -> scoreSingleNumber dice Die.Three
    | Fours          -> scoreSingleNumber dice Die.Four
    | Fives          -> scoreSingleNumber dice Die.Five
    | Sixes          -> scoreSingleNumber dice Die.Six
    | FullHouse      -> scoreFullHouse dice
    | FourOfAKind    -> scoreFourOfAKind dice
    | LittleStraight -> scoreLittleStraight dice
    | BigStraight    -> scoreBigStraight dice
    | Choice         -> scoreChoice dice
    | Yacht          -> scoreYacht dice
 

File: haskell/yacht/src/Yacht.hs
module Yacht (yacht, Category (..)) where

import Data.List (group, sort)

data Category
  = Ones
  | Twos
  | Threes
  | Fours
  | Fives
  | Sixes
  | FullHouse
  | FourOfAKind
  | LittleStraight
  | BigStraight
  | Choice
  | Yacht

yacht :: Category -> [Int] -> Int
yacht category dice =
  case category of
    Ones -> sumOfDice 1
    Twos -> sumOfDice 2
    Threes -> sumOfDice 3
    Fours -> sumOfDice 4
    Fives -> sumOfDice 5
    Sixes -> sumOfDice 6
    FullHouse
      | diceCounts == [2, 3] -> sum dice
      | otherwise -> 0
    FourOfAKind
      | diceCounts == [1, 4] -> 4 * head (tail sortedDice)
      | diceCounts == [5] -> 4 * head dice
      | otherwise -> 0
    LittleStraight
      | sortedDice == [1, 2, 3, 4, 5] -> 30
      | otherwise -> 0
    BigStraight
      | sortedDice == [2, 3, 4, 5, 6] -> 30
      | otherwise -> 0
    Choice -> sum dice
    Yacht
      | diceCounts == [5] -> 50
      | otherwise -> 0
  where
    sumOfDice die = sum $ filter (== die) dice
    sortedDice = sort dice
    diceCounts = sort $ map length $ group sortedDice
